### PR TITLE
Pair agent

### PR DESCRIPTION
**Pull request by torikulhabib**
Please correct it if there is something wrong.
I try to match the example and doc bluez as needed.
https://github.com/elementary/switchboard-plug-bluetooth/issues/30

### CODE REVIEW

File: src/DeviceRow.vala [hunk 1]
```diff
@@ -199,6 +199,7 @@ public class Bluetooth.DeviceRow : Gtk.ListBoxRow {
             var paired = changed.lookup_value ("Paired", new VariantType ("b"));
             if (paired != null) {
                 compute_status ();
+                device.trusted = device.paired; //paired an trusted
```
Comment (jeremypw) [inline src/DeviceRow.vala:202]: `//paired and trusted`

File: src/PairDialog.vala [hunk 2]
```diff
@@ -83,19 +107,66 @@ public class PairDialog : Granite.MessageDialog {
         }
 
         switch (auth_type) {
+            case AuthType.REQUESTPIN:
+                badge_icon = new ThemedIcon ("dialog-password");
+                secondary_text = _("Make sure the PIN on “%s” and the one you enter the PIN code below match.").printf (device_name);
+                var entry_pin = new Gtk.Entry ();
```
Comment (jeremypw) [inline src/PairDialog.vala:113]: Current recommended code style is to set properties on creation:
```
var entry_pin = new Gtk.Entry () {
    activates_default = true,
    xalign = 0.5f,
    ...
    width_chars = 16
};
```

File: src/PairDialog.vala [hunk 3]
```diff
@@ -83,19 +107,66 @@ public class PairDialog : Granite.MessageDialog {
...
+                custom_bin.margin_end = 25;
+                custom_bin.add (entry_pin);
+                custom_bin.show_all ();
+                var confirm_button = add_button (_("Confirm"), Gtk.ResponseType.ACCEPT);
+                confirm_button.get_style_context ().add_class (Gtk.STYLE_CLASS_SUGGESTED_ACTION);
+                break;
+            case AuthType.REQUESTPASSKEY:
+                badge_icon = new ThemedIcon ("dialog-password");
+                secondary_text = _("Make sure the Passkey on “%s” and the one you enter the Passkey code below match.").printf (device_name);
+                var entry_passkey = new Gtk.Entry ();
```
Comment (jeremypw) [inline src/PairDialog.vala:134]: Set properties on creation (see above)

File: src/PairDialog.vala [hunk 4]
```diff
@@ -115,23 +186,27 @@ public class PairDialog : Granite.MessageDialog {
...
             }
         });
-
-        ((DBusProxy)device).g_properties_changed.connect ((changed, invalid) => {
-            var paired = changed.lookup_value ("Paired", new VariantType ("b"));
-            if (paired != null && device.paired) {
-                destroy ();
-            }
+        destroy.connect (()=>{
+            device.blocked = false; // remove bloking
```
Comment (jeremypw) [inline src/PairDialog.vala:202]: `// remove blocking`

File: src/PairDialog.vala [hunk 5]
```diff
@@ -115,23 +186,27 @@ public class PairDialog : Granite.MessageDialog {
...
         });
     }
+
+    public string get_pincode () {
+        return pincodes;
+    }
+    public uint32 get_paskey () {
```
Comment (jeremypw) [inline src/PairDialog.vala:209]: Should be `get_passkey ()`?

**Review by torikulhabib — COMMENTED** [review]
I've changed it.

File: src/DeviceRow.vala [hunk 6]
```diff
@@ -186,20 +186,24 @@ public class Bluetooth.DeviceRow : Gtk.ListBoxRow {
 
         compute_status ();
         set_sensitive (adapter.powered);
-
-        (adapter as DBusProxy).g_properties_changed.connect ((changed, invalid) => {
+        // changed because to avoid :warning: Access to possible `null'. Perform a check or use an unsafe cast.:
```
Comment (davidmhewitt) [inline src/DeviceRow.vala:189]:
```suggestion
```

File: src/DeviceRow.vala [hunk 7]
```diff
@@ -186,20 +186,24 @@ public class Bluetooth.DeviceRow : Gtk.ListBoxRow {
...
+        ((DBusProxy) adapter).g_properties_changed.connect ((changed, invalid) => {
             var powered = changed.lookup_value ("Powered", new VariantType ("b"));
             if (powered != null) {
                 set_sensitive (adapter.powered);
                 this.changed ();
             }
         });
-
-        (device as DBusProxy).g_properties_changed.connect ((changed, invalid) => {
+        // changed because to avoid :warning: Access to possible `null'. Perform a check or use an unsafe cast.:
```
Comment (davidmhewitt) [inline src/DeviceRow.vala:197]:
```suggestion
```

File: src/PairDialog.vala [hunk 8]
```diff
@@ -26,6 +28,8 @@ public class PairDialog : Granite.MessageDialog {
     public ObjectPath object_path { get; construct; }
     public AuthType auth_type { get; construct; }
     public string passkey { get; construct; }
+    public string pincodes { get; construct; }
+    public uint32 passkey_uint32 { get; construct; }
```
Comment (davidmhewitt) [inline src/PairDialog.vala:32]:
```suggestion
    public string pincodes { get; private set; }
    public uint32 passkey_uint32 { get; private set; }
```

File: src/PairDialog.vala [hunk 9]
```diff
@@ -17,6 +17,8 @@
 
 public class PairDialog : Granite.MessageDialog {
     public enum AuthType {
+        REQUESTPIN,
+        REQUESTPASSKEY,
```
Comment (davidmhewitt) [inline src/PairDialog.vala:21]:
```suggestion
        REQUEST_PIN,
        REQUEST_PASSKEY,
```

File: src/PairDialog.vala [hunk 10]
```diff
@@ -37,6 +41,26 @@ public class PairDialog : Granite.MessageDialog {
         );
     }
 
+    public PairDialog.request_pin_code (ObjectPath object_path, Gtk.Window? main_window) {
+        Object (
+            auth_type: AuthType.REQUESTPIN,
```
Comment (davidmhewitt) [inline src/PairDialog.vala:46]:
```suggestion
            auth_type: AuthType.REQUEST_PIN,
```

File: src/PairDialog.vala [hunk 11]
```diff
@@ -37,6 +41,26 @@ public class PairDialog : Granite.MessageDialog {
...
+            buttons: Gtk.ButtonsType.CANCEL,
+            object_path: object_path,
+            primary_text: _("Enter Bluetooth PIN"),
+            transient_for: main_window
+        );
+    }
+
+    public PairDialog.request_passkey (ObjectPath object_path, Gtk.Window? main_window) {
+        Object (
+            auth_type: AuthType.REQUESTPASSKEY,
```
Comment (davidmhewitt) [inline src/PairDialog.vala:56]:
```suggestion
            auth_type: AuthType.REQUEST_PASSKEY,
```

File: src/PairDialog.vala [hunk 12]
```diff
@@ -83,19 +107,68 @@ public class PairDialog : Granite.MessageDialog {
         }
 
         switch (auth_type) {
+            case AuthType.REQUESTPIN:
```
Comment (davidmhewitt) [inline src/PairDialog.vala:110]:
```suggestion
            case AuthType.REQUEST_PIN:
```

File: src/PairDialog.vala [hunk 13]
```diff
@@ -83,19 +107,68 @@ public class PairDialog : Granite.MessageDialog {
...
+                badge_icon = new ThemedIcon ("dialog-password");
+                secondary_text = _("Make sure the PIN on “%s” and the one you enter the PIN code below match.").printf (device_name);
```
Comment (davidmhewitt) [inline src/PairDialog.vala:112]:
```suggestion
                secondary_text = _("Enter the pin displayed on “%s”").printf (device_name);
```

File: src/PairDialog.vala [hunk 14]
```diff
@@ -83,19 +107,68 @@ public class PairDialog : Granite.MessageDialog {
...
+                var entry_pin = new Gtk.Entry () {
+                    activates_default = true,
+                    xalign = 0.5f,
+                    input_hints = Gtk.InputHints.NO_SPELLCHECK | Gtk.InputHints.NONE,
+                    input_purpose = Gtk.InputPurpose.DIGITS,
+                    max_length = 16, // get from doc bluez
+                    width_chars = 16
+                };
+                entry_pin.get_style_context ().add_class (Granite.STYLE_CLASS_H1_LABEL);
+                entry_pin.changed.connect (()=>{
```
Comment (davidmhewitt) [inline src/PairDialog.vala:122]:
```suggestion
                entry_pin.changed.connect (() => {
```

File: src/PairDialog.vala [hunk 15]
```diff
@@ -83,19 +107,68 @@ public class PairDialog : Granite.MessageDialog {
...
+                    pincodes = entry_pin.text;
+                });
+                custom_bin.margin_start = 25;
+                custom_bin.margin_end = 25;
+                custom_bin.add (entry_pin);
+                custom_bin.show_all ();
+                var confirm_button = add_button (_("Confirm"), Gtk.ResponseType.ACCEPT);
+                confirm_button.get_style_context ().add_class (Gtk.STYLE_CLASS_SUGGESTED_ACTION);
+                break;
+            case AuthType.REQUESTPASSKEY:
```
Comment (davidmhewitt) [inline src/PairDialog.vala:132]:
```suggestion
            case AuthType.REQUEST_PASSKEY:
```

File: src/PairDialog.vala [hunk 16]
```diff
@@ -83,19 +107,68 @@ public class PairDialog : Granite.MessageDialog {
...
+                badge_icon = new ThemedIcon ("dialog-password");
+                secondary_text = _("Make sure the Passkey on “%s” and the one you enter the Passkey code below match.").printf (device_name);
```
Comment (davidmhewitt) [inline src/PairDialog.vala:134]:
```suggestion
                secondary_text = _("Enter the passkey displayed on “%s”").printf (device_name);
```

File: src/PairDialog.vala [hunk 17]
```diff
@@ -83,19 +107,68 @@ public class PairDialog : Granite.MessageDialog {
...
+                var entry_passkey = new Gtk.Entry () {
+                    activates_default = true,
+                    xalign = 0.5f,
+                    input_hints = Gtk.InputHints.NONE,
+                    input_purpose = Gtk.InputPurpose.NUMBER,
+                    max_length = 6, //get from doc bluez
+                    width_chars = 6
+                };
+                entry_passkey.get_style_context ().add_class (Granite.STYLE_CLASS_H1_LABEL);
+                entry_passkey.changed.connect (()=>{
```
Comment (davidmhewitt) [inline src/PairDialog.vala:144]:
```suggestion
                entry_passkey.changed.connect (() => {
```

File: src/PairDialog.vala [hunk 18]
```diff
@@ -115,23 +188,27 @@ public class PairDialog : Granite.MessageDialog {
...
                     break;
             }
         });
-
-        ((DBusProxy)device).g_properties_changed.connect ((changed, invalid) => {
-            var paired = changed.lookup_value ("Paired", new VariantType ("b"));
-            if (paired != null && device.paired) {
-                destroy ();
-            }
+        destroy.connect (()=>{
```
Comment (davidmhewitt) [inline src/PairDialog.vala:203]:
```suggestion
        destroy.connect (() => {
```

File: src/PairDialog.vala [hunk 19]
```diff
@@ -115,23 +188,27 @@ public class PairDialog : Granite.MessageDialog {
...
+            device.blocked = false; // remove blocking
         });
     }
+
+    public string get_pincode () {
+        return pincodes;
+    }
+    public uint32 get_upasskey () {
```
Comment (davidmhewitt) [inline src/PairDialog.vala:211]:
```suggestion
    }
    
    public uint32 get_upasskey () {
```

**Review by davidmhewitt — CHANGES_REQUESTED** [review]

File: src/PairDialog.vala [hunk 20]
```diff
@@ -26,6 +28,8 @@ public class PairDialog : Granite.MessageDialog {
     public ObjectPath object_path { get; construct; }
     public AuthType auth_type { get; construct; }
     public string passkey { get; construct; }
+    public string pincodes { get; private set; }
```
Comment (davidmhewitt) [inline src/PairDialog.vala:31]:
```suggestion
    public string pincode { get; private set; }
```

File: src/PairDialog.vala [hunk 21]
```diff
@@ -83,19 +107,68 @@ public class PairDialog : Granite.MessageDialog {
...
+                    activates_default = true,
+                    xalign = 0.5f,
+                    input_hints = Gtk.InputHints.NO_SPELLCHECK | Gtk.InputHints.NONE,
+                    input_purpose = Gtk.InputPurpose.DIGITS,
+                    max_length = 16, // get from doc bluez
+                    width_chars = 16
+                };
+                entry_pin.get_style_context ().add_class (Granite.STYLE_CLASS_H1_LABEL);
+                entry_pin.changed.connect (() => {
+                    pincodes = entry_pin.text;
```
Comment (davidmhewitt) [inline src/PairDialog.vala:123]:
```suggestion
                    pincode = entry_pin.text;
```

File: src/PairDialog.vala [hunk 22]
```diff
@@ -115,23 +188,28 @@ public class PairDialog : Granite.MessageDialog {
...
         });
     }
+
+    public string get_pincode () {
+        return pincodes;
+    }
+    
+    public uint32 get_upasskey () {
+        return passkey_uint32;
+    }
```
Comment (davidmhewitt) [inline src/PairDialog.vala:214]:
```suggestion
```

File: src/Services/Agent.vala [hunk 23]
```diff
@@ -27,70 +28,76 @@ public errordomain BluezError {
...
-        unregistered ();
     }
 
     public string request_pin_code (ObjectPath device) throws Error, BluezError {
-        return "";
+        pair_dialog = new PairDialog.request_pin_code (device, main_window);
+        pair_dialog.present ();
+        pair_dialog.run ();
+        pair_dialog.destroy ();
+        return pair_dialog.get_pincode ();
```
Comment (davidmhewitt) [inline src/Services/Agent.vala:56]:
```suggestion
        return pair_dialog.pincode;
```

File: src/PairDialog.vala [hunk 24]
```diff
@@ -83,19 +107,68 @@ public class PairDialog : Granite.MessageDialog {
...
+                    activates_default = true,
+                    xalign = 0.5f,
+                    input_hints = Gtk.InputHints.NONE,
+                    input_purpose = Gtk.InputPurpose.NUMBER,
+                    max_length = 6, //get from doc bluez
+                    width_chars = 6
+                };
+                entry_passkey.get_style_context ().add_class (Granite.STYLE_CLASS_H1_LABEL);
+                entry_passkey.changed.connect (() => {
+                    passkey_uint32 = (uint32) uint64.parse (entry_passkey.text);
```
Comment (davidmhewitt) [inline src/PairDialog.vala:145]:
```suggestion
                    passkey_uint32 = uint.parse (entry_passkey.text);
```

**Review by davidmhewitt — CHANGES_REQUESTED** [review]

File: src/Services/Agent.vala [hunk 25]
```diff
@@ -27,70 +28,76 @@ public errordomain BluezError {
...
+        pair_dialog.destroy ();
     }
 
     public uint32 request_passkey (ObjectPath device) throws Error, BluezError {
-        return 0;
+        pair_dialog = new PairDialog.request_passkey (device, main_window);
+        pair_dialog.present ();
+        pair_dialog.run ();
+        pair_dialog.destroy ();
+        return pair_dialog.get_upasskey ();
```
Comment (davidmhewitt) [inline src/Services/Agent.vala:71]:
```suggestion
        return pair_dialog.passkey_uint32;
```

**Review by jeremypw — COMMENTED** [review]
Needs conflicts fixing.
Could successfully pair and connect with Android phone entering a passcode but the same dialog appeared whether or not the pairing was initiated from elementary or the phone (that is, the "request" dialog and not the "display" dialog.  Is this correct?

Still paired connected OK with headphones not requiring authentication.

Will check behaviour against Ubuntu.  Needs testing by others with different device authentication modes.

**Pull request closed without merging**